Zend/zend_exceptions.c

```diff
@@ -254,41 +254,45 @@ ZEND_API void zend_clear_exception(void) /* {{{ */
 }
 /* }}} */
 
+/* Same as OBJ_PROP_NUM(), but checks the offset is correct when Zend is built in debug mode. */
+static zend_always_inline zval *zend_obj_prop_num_checked(zend_object *object, uint32_t prop_num, zend_string *str)
+{
+#if ZEND_DEBUG
+	zend_class_entry *old_scope = EG(fake_scope);
+	EG(fake_scope) = i_get_exception_base(object);
+	const zend_property_info *prop_info = zend_get_property_info(object->ce, str, true);
+	ZEND_ASSERT(OBJ_PROP_TO_NUM(prop_info->offset) == prop_num);
```

Comment:
I don't think you can depend on this since https://github.com/php/php-src/pull/17870. These properties are not final, so a child class may add hooks to them. A simple solution would be to check for `ce->num_hooked_props == 0` and use a fallback to `zend_update_property_ex()` otherwise.

Reply:
It's fine for the private properties though.

Reply:
Indeed, that would be fine. Not all properties are here though.

Reply:
Ugh, right, so that has likely also broken the same optimization I did in DOM for some props a long time ago. Since introducing hooks in a subclass of an exception type is likely uncommon, I'll just add the `num_hooked_props` check I guess and figure it out from there.